projects/fluid_simulation/sim.cpp

```diff
@@ -0,0 +1,435 @@
+    bci.usageFlags.bits.uniformBuffer = true;
+    bci.memoryUsage                   = ppx::grfx::MEMORY_USAGE_CPU_TO_GPU;
+    PPX_CHECKED_CALL(GetApp()->GetDevice()->CreateBuffer(&bci, &mGraphics.mTransformBuffer));
+
+    ppx::float4x4 mat   = glm::rotate(0.0f, ppx::float3(0, 0, 1));
+    void*         pData = nullptr;
+    PPX_CHECKED_CALL(mGraphics.mTransformBuffer->MapMemory(0, &pData));
+    memcpy(pData, &mat, sizeof(mat));
+    mGraphics.mTransformBuffer->UnmapMemory();
+
+    ppx::grfx::PipelineInterfaceCreateInfo pici = {};
+    pici.setCount                               = 1;
+    pici.sets[0].set                            = 0;
+    pici.sets[0].pLayout                        = mGraphics.mDescriptorSetLayout;
+    PPX_CHECKED_CALL(GetApp()->GetDevice()->CreatePipelineInterface(&pici, &mGraphics.mPipelineInterface));
+
+    mDraw = std::make_unique<GraphicShader>(this);
+}
+
+ppx::int2 FluidSimulation::GetResolution(float resolution)
+{
+    float aspectRatio = static_cast<float>(GetApp()->GetWindowWidth()) / static_cast<float>(GetApp()->GetWindowHeight());
+    if (aspectRatio < 1.0) {
+        aspectRatio = 1.0 / aspectRatio;
+    }
+
+    int min = round(resolution);
+    int max = round(resolution * aspectRatio);
+
+    return (GetApp()->GetWindowWidth() > GetApp()->GetWindowHeight()) ? ppx::int2(max, min) : ppx::int2(min, max);
+}
+
+void FluidSimulation::InitTextures()
+{
+    ppx::int2 simRes = GetResolution(mConfig.simResolution);
+    ppx::int2 dyeRes = GetResolution(mConfig.dyeResolution);
+
+    // Generate all the textures.
+    mCheckerboardTexture = std::make_unique<Texture>(this, "checkerboard", GetApp()->GetWindowWidth(), GetApp()->GetWindowHeight(), GetApp()->GetSwapchain()->GetColorFormat());
+    mDisplayTexture      = std::make_unique<Texture>(this, "display", GetApp()->GetWindowWidth(), GetApp()->GetWindowHeight(), GetApp()->GetSwapchain()->GetColorFormat());
+    mDitheringTexture    = std::make_unique<Texture>(this, "fluid_simulation/textures/LDR_LLL1_0.png");
+    mDrawColorTexture    = std::make_unique<Texture>(this, "draw color", GetApp()->GetWindowWidth(), GetApp()->GetWindowHeight(), GetApp()->GetSwapchain()->GetColorFormat());
+    mDyeTexture[0]       = std::make_unique<Texture>(this, "dye[0]", dyeRes.x, dyeRes.y, kRGBA);
+    mDyeTexture[1]       = std::make_unique<Texture>(this, "dye[1]", dyeRes.x, dyeRes.y, kRGBA);
+    mVelocityTexture[0]  = std::make_unique<Texture>(this, "velocity[0]", simRes.x, simRes.y, kRG);
+    mVelocityTexture[1]  = std::make_unique<Texture>(this, "velocity[1]", simRes.x, simRes.y, kRG);
+
+    InitBloomTextures();
+    InitSunraysTextures();
+}
+
+void FluidSimulation::InitBloomTextures()
+{
+    ppx::int2 res = GetResolution(mConfig.bloomResolution);
+    mBloomTexture = std::make_unique<Texture>(this, "bloom", res.x, res.y, kRGBA);
+    PPX_ASSERT_MSG(mBloomTextures.empty(), "Bloom FBOs already initialized");
```

Comment:
`FBOs` is OpenGL parlance. `textures` is better here

Reply:
Yup.